docs/rules/no-cycle.md

```diff
@@ -0,0 +1,63 @@
+# import/no-cycle
+
+Ensures that there is no resolvable path back to this module via its dependencies.
+
+This includes cycles of depth 1 (imported module imports me) to `Infinity`, if the
+[`maxDepth`](#maxdepth) option is not set.
+
+```js
+// dep-b.js
+import './dep-a.js'
+
+export function b() { /* ... */ }
+
+// dep-a.js
+import { b } from './dep-b.js' // reported: Dependency cycle detected.
+```
+
+This rule does _not_ detect imports that resolve directly to the linted module;
+for that, see [`no-self-import`].
+
+
+## Rule Details
+
+### Options
+
+By default, this rule only detects cycles for ES6 imports, but see the [`no-unresolved` options](./no-unresolved.md#options) as this rule also supports the same `commonjs` and `amd` flags. However, these flags only impact which import types are _linted_; the
+import/export infrastructure only registers `import` statements in dependencies, so
```

Comment:
It seems super important that this can detect requires as well as imports. What’s involved in making that happen?

Reply:
Yeah, I knew you'd pick up on that as I realized it and wrote it out. 😅 

Right now, all the normal static analysis inspection of dependencies only picks up imports/exports from the topmost scope of the module. 

Could also potentially nab any single-argument `require`s that are either alone or in `AssignmentExpression`s but everything gets dicier once `require`s are in the mix, since they could be present in any scope in the module, technically.

I want to avoid blocking merging and shipping this on supporting deep registration of CJS `require`s, though.

Reply:
Could you use an ast selector like `CallExpression[callee.name=“require”]` for the visitor?

Reply:
For sure. That's certainly what the `moduleVisitor` implementation does to find and lint `require`s.

It would be a substantial improvement, and I think I am probably overstating the complexity. I'll try to take a second look.